spmd/compiler/fusion.py

```diff
@@ -162,22 +214,448 @@ def _scan_graph_for_fusion_elements(
+    gi: GraphInfo, gm: fx.GraphModule, in_fe_list: list[FusionElement]
+) -> None:
+    """first half of fusion - move desired items to buffer and create graph"""
+    buffer_node = gi.global_buffer_node
+    buffer_size = gi.global_buffer_size
+
+    copy_list = in_fe_list
+
+    num_fusion_elements = len(copy_list)
+
+    def copy_to_buffer(
+        buffer: torch.Tensor, tensor_list: List[torch.Tensor]
+    ) -> torch.Tensor:
+        offset = 0
+        for t in tensor_list:
+            size = t.numel()
+            buffer[offset : offset + size] = t.view(-1)
+            offset += size
+        return buffer
+
+    # setup dummy vars
+    buffer = None
+    if gi.tracing_buffer is None:
+        buffer = torch.empty(buffer_size)
+        gi.tracing_buffer = buffer
+    elif gi.tracing_buffer:
+        buffer = buffer = torch.empty(buffer_size)
+
+    tlist = []
+    for item in copy_list:
+        a = torch.zeros((item.shape[0], item.shape[1]))  # type: ignore
+        tlist.append(a)
+
+    load_gm = make_fx(copy_to_buffer)(buffer, tlist)
+
+    subnodemap = create_graph_node_map(load_gm)
+
+    # update load loop to use main graph items
+    fn_list = []
+    pl_list = []
+    for node in load_gm.graph.nodes:
+        if node.op == OP.PLACEHOLDER:
+            pl_list.append(node)
+        elif node.op == OP.CALL_FUNCTION:
+            fn_list.append(node)
+
+    # create placeholder remapping
+    pl_map: Dict[fx.Node, fx.Node] = {}
+    pl_map[pl_list[0]] = gi.global_buffer_node  # type: ignore
+
+    for i in range(num_fusion_elements):
+        pl_map[pl_list[i + 1]] = in_fe_list[i].grad_tensor_node  # type: ignore
+
+    insert_node = in_fe_list[-1].comm_node
+
+    def remap_copy_args(in_node: fx.Node) -> fx.Node:
+        out_node = in_node
+        if in_node in pl_map:
+            out_node = pl_map[in_node]  # type: ignore
+        elif in_node in value_remap:
+            out_node = value_remap[in_node]
+        return out_node
+
+    value_remap = {}
+    with gm.graph.inserting_before(insert_node):
+        for innernode in load_gm.graph.nodes:
+            if innernode.op in [OP.PLACEHOLDER, OP.OUTPUT]:
+                continue
+            value_remap[innernode] = gm.graph.node_copy(
+                innernode, remap_copy_args
+            )
+
+    _update_new_copy_nodes_users(value_remap)
+
+    # update allreduce to use buffer
+    # (we currently don't) have to make our own all_reduce/comm_wait section
+    # # TODO - pg group matching
+    # _build_buffer_comm_graph(gm, gi)
+
+    buffer_comm_node = in_fe_list[-1].comm_node
+    buffer_comm_node.update_arg(0, [buffer_node])  # type: ignore
+
+
+def _build_buffer_comm_graph(
+    gi: GraphInfo, gm: fx.GraphModule
+) -> fx.GraphModule:
+    """have to make our own all_reduce and wait subgraph for buffer"""
+    # from torch.distributed._spmd.comm_tensor import CommTensor
+    from torch.distributed.distributed_c10d import (  # ProcessGroup,; Work,; all_reduce,
```

Comment:
import at the top? or is this a circular issue

Reply:
atm this is a stub function so I kept the includes local to it rather than at the top.  
Also I see black has scrunched the commented out aspects in 306 above and have removed.
I've also added a comment in the doc string that this is only a stub for clarity.